### ISSUE TITLE
(Data loss) Issue with symlink causes entire remote IMAP store to be purged/deleted.

### ISSUE THREAD

**Issue by Dev0Null** [OfflineIMAP/offlineimap#426]
#### General informations

- system/distribution (with version): Ubuntu 4.4.0-59-generic x86_64
- offlineimap version (`offlineimap -V`): 6.6.1
- Python version:2.7.12
- server name or domain: 
- CLI options: none

#### Configuration file offlineimaprc

```
REMOVE PRIVATE DATA.
```

#### pythonfile (if any)

```
REMOVE PRIVATE DATA.
```

#### Logs, error

```
REMOVE PRIVATE DATA.
```

#### Steps to reproduce the error

- Trivial rc file setup (`https://github.com/connermcd/dotfiles/blob/master/.offlineimaprc.template`).
- Create a symlink to `~/.offlineimap`.
- Sync emails to local store via `offlineimap` (works).
- Invalidate the symlink to `~/.offlineimap` (ie: have it point to an empty folder).
- Run `offlineimap`.
- Result: `offlineimap` assumes the local store is empty, rather than invalid, and purges all e-mails from the remote store (not fun).
- Suggestions: a validation/sanity check routine should be executed to prevent this.

**Comment by nicolas33** [comment]
Deleting the whole content of a folder should only happen while deleting the local folder. We already handle it right when the content of folder is cleared (resync) but not when the parent folder is likely deleted.

**Comment by nicolas33** [comment]
I did not read the report right the first time. `~/.offlineimap` is the metadata path. Do you pretend that clearing the metadata path makes offlineimap delete the remote emails?

**Comment by Dev0Null** [comment]
Mistake on my part in initial report. Should read "create a symlink to `~/Maildir`. When the symlink to maildir points to an empty folder, the issue occurs.

**Comment by chris001** [comment]
@Dev0Null  In case you didn't know, you can edit your initial report and fix the mistake, just click on the "pencil" icon on the top right, near the "reaction" icon.

**Comment by Dev0Null** [comment]
@chris001 I know. I didn't want to confuse future readers by making the conversation above moot. Cheers.

**Comment by chris001** [comment]
This issue brings attention to the sync algorithm.  Originally it was designed so the local is always the master in the sync process, with no chance to undo a real delete.

Should the local Maildir (or local IMAP) always be the absolute master in the sync?

What algorithm does apple use to sync iphone/ipad emails (equivalent to our local maildir or local imap) to icloud (equivalent to the remote mail server) with hardly any reports of total loss of all email?

Same for google android and gmail, what's their sync algorithm handle local/remote emails, when one side is missing a small or large number of messages?  It doesn't instantly delete the missing message from the remote side permanently.

A simple way to handle this: don't really delete mail, just move it to a holding area/recycle bin/soft-delete folder tree, for X days, default 30 or 60, 90, 180, 365 days, after which time, the sync algorithm can be configured to purge/delete it permanently.

**Comment by Dev0Null** [comment]
@chris001 Updated bug report. Purged now-outdated comments.

**Comment by nicolas33** [comment]
Please no! Don't edit the comments afterwards except for minor fixes to avoid breaking the next comments.

AFAICT, we must only clear a folder when the local directory is still there and empty. We fail at this.

**Comment by nicolas33** [comment]
> Originally it was designed so the local is always the master,

No, there is no concept of master. Sync in offlineimap means propagate changes to the other sides. On conflicts the result is "keep the data".

> for deleted messages, in the sync process, with no chance to undo a real delete. When the user deletes the local copy of the email message, on the next sync it'll really delete the remote copy of the email message from the remote server also.

Yes.

> What algorithm does apple use to sync iphone/ipad emails (equivalent to our local maildir or local imap) to icloud (equivalent to the remote mail server) with hardly any reports of total loss of all email?

I don't use above tools. I'd say they don't sync but use standard IMAP (perhaps with disconnected mode).

> A simple way to handle this: don't really delete mail, just move it to a holding area/recycle bin/soft-delete folder tree, for X days, default 30 or 60, 90, 180, 365 days, after which time, the sync algorithm can be configured to purge/delete it permanently.

This could be a new (optional) feature. You might want to open a new issue about that.

**Comment by nicolas33** [comment]
This behaviour is now documented.

**Comment by mowny** [comment]
@chris001 Clearing the messages on an iphone also clears them from the iCloud account. I was called to fix such a problem last year and luckily there was a notebook with pop3 so only the 3 weeks from the last pop3 download up to the deletion where gone. Was told there were several warnings, though, which he chose to ignore.

Suggestions:
- require "--force" or something when there are more than N messages to delete
- require some switch to delete anything at all, just add (to) folders otherwise

(Right now I only have jessie's 6.3.4-1 here so I can't check whether something like that is already implemented in 7.0.12)
